src/Sparrow.Server/TimestampAccessor.cs

```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Threading;
+
+namespace Sparrow.Server
+{
+    // The idea behind the timestamp accessor is to be able to query the UTC datetime value
+    // while at the same time avoid multiple calls to the kernel operations when they are not
+    // needed for accuracy. 
+    public static class TimestampAccessor
+    {
+        private static long _timestamp;
+        private static readonly Timer _timer;
+
+        // Static constructor to initialize the timestamp and timer
+        static TimestampAccessor()
+        {
+            _timestamp = DateTime.UtcNow.Ticks;
```

Comment:
This is implemented using VDSO, basically it is reading a known memory location.
On Linux, `UtcNow` is getting from:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/DateTime.Unix.cs#L14
https://github.com/dotnet/runtime/blob/5f8a2ee126eb1b4c96d867efa6b7914839b29910/src/native/libs/System.Native/pal_datetime.c#L32

That should invoke the VDSO path, meaning that the actual call is literally reading a memory location.

On Windows, this goes to:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs

That is using KUSER_SHARED_DATA, basically the same concept as VDSO, but without the need for code injection, since that is implemented directly by kernel32.

See more here: 
https://osm.hpi.de/wrk/2007/08/getting-os-information-the-kuser_shared_data-structure/

Kestrel is doing something similar:
https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Internal/Http/DateHeaderValueManager.cs

But on a 1 sec scale (also formatting the date itself).


In other words, this looks very much like you are just ending up copying the value to your own memory, whereas we'll call that anyway through the normal API.

Reply:
True, but one memory is in the kernel and requires an interop call which costs a lot. I was able to get rid completely of the cost after doing this, which means that the overhead is actually quite high.

Reply:
There is no kernel/usermode switch here, mind.